programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+        )?;
+        validate!(
+            new_constituent_correlations.len() == n,
+            ErrorCode::DefaultError,
+            "new_corrs length must equal number of number of other constituents ({})",
+            n
+        )?;
+        for &c in new_constituent_correlations {
+            validate!(
+                c <= PERCENTAGE_PRECISION_I64,
+                ErrorCode::DefaultError,
+                "correlation must be ≤ PERCENTAGE_PRECISION"
+            )?;
+        }
+
+        let new_n = n + 1;
+        let mut buf = Vec::with_capacity(new_n * new_n);
+
+        for i in 0..n {
+            buf.extend_from_slice(&self.correlations[i * n..i * n + n]);
+            buf.push(new_constituent_correlations[i]);
+        }
+
+        buf.extend_from_slice(new_constituent_correlations);
+        buf.push(PERCENTAGE_PRECISION_I64);
+
+        self.correlations = buf;
+
+        debug_assert_eq!(self.correlations.len(), new_n * new_n);
+
+        Ok(())
+    }
+
+    pub fn set_correlation(&mut self, i: u16, j: u16, corr: i64) -> DriftResult {
+        let num_constituents = (self.correlations.len() as f64).sqrt() as usize;
+        validate!(
+            i < num_constituents as u16,
+            ErrorCode::InvalidConstituent,
+            "Invalid constituent_index i = {}, ConstituentCorrelation len = {}",
+            i,
+            num_constituents
+        )?;
+        validate!(
+            j < num_constituents as u16,
+            ErrorCode::InvalidConstituent,
+            "Invalid constituent_index j = {}, ConstituentCorrelation len = {}",
+            j,
+            num_constituents
+        )?;
+        validate!(
+            corr <= PERCENTAGE_PRECISION_I64,
+            ErrorCode::DefaultError,
+            "ConstituentCorrelation correlations must be between 0 and PERCENTAGE_PRECISION"
+        )?;
+
+        self.correlations[(i as usize * num_constituents + j as usize) as usize] = corr;
+        self.correlations[(j as usize * num_constituents + i as usize) as usize] = corr;
+
+        self.validate()?;
+
+        Ok(())
+    }
+}
+
+impl<'a> AccountZeroCopy<'a, i64, ConstituentCorrelationsFixed> {
+    pub fn get_correlation(&self, i: u16, j: u16) -> DriftResult<i64> {
+        let num_constituents = (self.len() as f64).sqrt() as usize;
+        validate!(
+            i < num_constituents as u16,
+            ErrorCode::InvalidConstituent,
+            "Invalid constituent_index i = {}, ConstituentCorrelation len = {}",
+            i,
+            num_constituents
+        )?;
+        validate!(
+            j < num_constituents as u16,
+            ErrorCode::InvalidConstituent,
+            "Invalid constituent_index j = {}, ConstituentCorrelation len = {}",
+            j,
+            num_constituents
+        )?;
+
+        let corr = self.get((i as usize * num_constituents + j as usize) as u32);
+        Ok(*corr)
+    }
+}
+
+pub fn get_gamma_covar_matrix(
```

Comment:
did we think about just pre-computing this whenever we up correlation/gamma/volatility?